pkg/netns/netns.go

```diff
@@ -4,88 +4,302 @@
 package netns
 
 import (
-	"errors"
 	"fmt"
 	"os"
-	"os/exec"
+	"path"
 	"path/filepath"
+	"runtime"
 	"strings"
+	"sync"
 
-	"github.com/containernetworking/plugins/pkg/ns"
-	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+
+	// TODO consider dropping this dep entirely - it is only used here now
+	"github.com/vishvananda/netns"
 )
 
-// RemoveIfFromNetNSIfExists removes the given interface from the given
-// network namespace.
+const bindMountPath = "/run/netns" /* Bind mount path for named netns */
+
+type netNSHandle struct {
+	handle int
+	Name   string
+}
+
+type NetNsHandle interface {
+	// Returns a file descriptor representing this object's network namespace
+	Fd() int
+
+	// Cleans up this instance of the network namespace; if this instance
+	// is the last user the namespace will be destroyed. Will Unpin first if
+	// instance is currently Pin-ed.
+	Close() error
+
+	// "Pins" or bind-mounts the netns created here to a specific path in nsfs.
```

Comment:
_If_ (big IF) we do need pins, I would at least like to be opinionated about where pinned NSes created by us live, since I doubt there's any need for us to manage or handle netnses we don't create, and we would probably prefer people not do that.